File: Java/Sprintf.java

import java.io.*;

public class Sprintf {

	private static void println(String data) throws IOException{
		System.out.println(data);
	}

	public static void main(String[] args) throws IOException{

		println("これから文字列フォーマットを試します");

		println("整数");
		println("String.format(\"%d\",255) = \""+(   String.format("%d",255)   )+"\" (DEC)");
		println("String.format(\"%o\",255) = \""+(   String.format("%o",255)   )+"\" (OCT)");
		println("String.format(\"%x\",255) = \""+(   String.format("%x",255)   )+"\" (HEX)");
		println("String.format(\"%X\",255) = \""+(   String.format("%X",255)   )+"\" (HEX)");

		println("\r\n実数");
		println("String.format(\"%f\",0.018) = \""+(   String.format("%f",0.018)   )+"\"");
		println("String.format(\"%g\",0.018) = \""+(   String.format("%g",0.018)   )+"\"");
		println("String.format(\"%e\",0.018) = \""+(   String.format("%e",0.018)   )+"\"");
		println("String.format(\"%E\",0.018) = \""+(   String.format("%E",0.018)   )+"\"");

		println("\r\n真偽値");
		println("String.format(\"%b\",true) = \""+(   String.format("%b",true)   )+"\"");

		println("\r\n文字");
		println("String.format(\"%c\",115)   = \""+(     String.format("%c",115)     )+"\"");
		println("String.format(\"%s\",\"abc\") = \""+(   String.format("%s","abc")   )+"\"");
			// %s は任意のオブジェクトで .toString() の結果を代入する

		println("\r\n基数の表記");
		println("String.format(\"%#o\",255) = \""+(   String.format("%#o",255)   )+"\"");
		println("String.format(\"%#x\",255) = \""+(   String.format("%#x",255)   )+"\"");
		println("String.format(\"%#X\",255) = \""+(   String.format("%#X",255)   )+"\"");

		println("\r\n正負の表記");
		println("String.format( \"%d\",+18) = \""+(   String.format( "%d",+18)   )+"\"");
		println("String.format(\"%+d\",+18) = \""+(   String.format("%+d",+18)   )+"\"");
		println("String.format(\"% d\",+18) = \""+(   String.format("% d",+18)   )+"\"");
		println("String.format( \"%d\",-18) = \""+(   String.format( "%d",-18)   )+"\"");
		println("String.format(\"%+d\",-18) = \""+(   String.format("%+d",-18)   )+"\"");
		println("String.format(\"% d\",-18) = \""+(   String.format("% d",-18)   )+"\"");

		println("\r\n精度");
		println("String.format(\"%.3f\",0.018) = \""+(     String.format("%.3f",0.018)   )+"\" (f: 小数点以下の桁数)");
		println("String.format(\"%.3e\",0.018) = \""+(     String.format("%.3e",0.018)   )+"\" (e,g: 有効桁数-1)");
		println("String.format(\"%.2s\",\"abc\") = \""+(   String.format("%.2s","abc")   )+"\" (s: 文字数)");

		println("\r\n幅");
		println("String.format(\"%10d\",18)     = \""+(     String.format("%10d",18)      )+"\"");
		println("String.format(\"%10f\",0.018)  = \""+(     String.format("%10f",0.018)   )+"\"");
		println("String.format(\"%10e\",0.018)  = \""+(     String.format("%10e",0.018)   )+"\"");
		println("String.format(\"%10s\",\"ab\")   = \""+(   String.format("%10s","ab")    )+"\"");

		println("\r\n0埋め");
		println("String.format(\"%010d\",18)    = \""+(     String.format("%010d",18)      )+"\"");
		println("String.format(\"%010f\",0.018) = \""+(     String.format("%010f",0.018)   )+"\"");
		println("String.format(\"%010e\",0.018) = \""+(     String.format("%010e",0.018)   )+"\"");

		println("\r\n桁区切り");
		println("String.format(\"%,d\",16777216)  = \""+(   String.format("%,d",16777216)   )+"\"");
		println("String.format(\"%,g\",167772.16) = \""+(   String.format("%,g",167772.16)   )+"\"");

		println("\r\n左寄せ");
		println("String.format( \"%10d\",18) = \""+(   String.format("%10d",18)    )+"\"");
		println("String.format(\"%-10d\",18) = \""+(   String.format("%-10d",18)   )+"\"");

		println("\r\n組合せ");
		println("String.format(\"%d年%d月%d日\",2007,7,14) = \""+(   String.format("%d年%d月%d日",2007,7,14)   )+"\"");
		println("String.format(\"x = %+08.2f\",0.375)      = \""+(   String.format("x = %+08.2f",0.375)      )+"\"");

		println("\r\n引数並び");
		println("String.format(\"%3$d年%1$d月%2$d日\",7,14,2007) = \""+(   String.format("%3$d年%1$d月%2$d日",7,14,2007)   )+"\"");
		println("String.format(\"%2$sは%2$s。%1$sは%1$s。\",\"あれ\",\"これ\") = \""+(   String.format("%2$sは%2$s。%1$sは%1$s。","あれ","これ")   )+"\"");

		println("\r\n");

	}
}

File: C/Sprintf.c
#include <stdio.h>
#include "Header.h"

void Sprintf(){

	printLn2(1,"これから文字列フォーマットを試します",1);

	printLn2(1,"整数",1);
	printf("sprintf(\"%%d\",255) = %d (整数)",255);nr(1); // int,short
	printf("sprintf(\"%%i\",255) = %i (整数)",255);nr(1);
	printf("sprintf(\"%%u\",255) = %u (符号なし整数)",255);nr(1); // unsigned int,unsigned short
	printf("sprintf(\"%%o\",255) = %o (OCT)",255);nr(1); // int,short,unsigned int,unsigned short
	printf("sprintf(\"%%x\",255) = %x (HEX)",255);nr(1); // int,short,unsigned int,unsigned short
	printf("sprintf(\"%%X\",255) = %X (HEX)",255);nr(1); // int,short,unsigned int,unsigned short
	/*
		long, unsigned long 型の整数で上のような演算を行いたい場合は,それぞれ指定子の前に l を付加する。
		e.g. %d → %ld
		long long の場合は %lld のようにする
	*/

	printLn2(1,"実数",1);
	printf("sprintf(\"%%f\",0.018) = \"%f\"",0.018);nr(1);
	printf("sprintf(\"%%g\",0.018) = \"%g\"",0.018);nr(1);
	printf("sprintf(\"%%e\",0.018) = \"%e\"",0.018);nr(1);
	printf("sprintf(\"%%E\",0.018) = \"%E\"",0.018);nr(1);
	/*
		double 型の整数で上のような演算を行いたい場合は,それぞれ指定子の前に l を付加する。
		...というか,今時は浮動小数の演算は基本的に double なので,それが普通である。
		e.g. %f → %lf
		long double の場合は %Lf とする
	*/

	printLn2(1,"文字",1);
	printf("sprintf(\"%%c\",115)   = \"%c\"",115);nr(1); // char
	printf("sprintf(\"%%s\",\"abc\") = \"%s\"","abc");nr(1); // char *

	printLn2(1,"基数の表記",1);
	printf("sprintf(\"%%#o\",255) = \"%#o\"",255);nr(1);
	printf("sprintf(\"%%#x\",255) = \"%#x\"",255);nr(1);
	printf("sprintf(\"%%#x\",255) = \"%#X\"",255);nr(1);

	printLn2(1,"正負の表記",1);
	printf("sprintf( \"%%d\",+18) = \"%d\"",+18);nr(1);
	printf("sprintf(\"%%+d\",+18) = \"%+d\"",+18);nr(1);
	printf("sprintf(\"%% d\",+18) = \"% d\"",+18);nr(1);
	printf("sprintf( \"%%d\",-18) = \"%d\"",-18);nr(1);
	printf("sprintf(\"%%+d\",-18) = \"%+d\"",-18);nr(1);
	printf("sprintf(\"%% d\",-18) = \"% d\"",-18);nr(1);

	printLn2(1,"精度",1);
	printf("sprintf(\"%%.3d\",18)    = \"%.3d\"",18);nr(1);      // d,i,u,o,x: 整数桁数
	printf("sprintf(\"%%.3f\",0.018) = \"%.3f\"",0.018);nr(1);   // f: 小数点以下の桁数
	printf("sprintf(\"%%.3e\",0.018) = \"%.3e\"",0.018);nr(1);   // e: 小数点以下の有効桁数
	printf("sprintf(\"%%.2s\",\"abc\") = \"%.2s\"","abc");nr(1); // s: 文字数

	printLn2(1,"幅",1);
	printf("sprintf(\"%%10d\",18)    = \"%10d\"",18);nr(1);
	printf("sprintf(\"%%10f\",0.018) = \"%10f\"",0.018);nr(1);
	printf("sprintf(\"%%10e\",0.018) = \"%10e\"",0.018);nr(1);
	printf("sprintf(\"%%10s\",\"ab\")  = \"%10s\"","ab");nr(1);

	printLn2(1,"0埋め",1);
	printf("sprintf(\"%%010d\",18)    = \"%010d\"",18);nr(1);
	printf("sprintf(\"%%010f\",0.018) = \"%010f\"",0.018);nr(1);
	printf("sprintf(\"%%010e\",0.018) = \"%010e\"",0.018);nr(1);

	printLn2(1,"左寄せ",1);
	printf("sprintf( \"%%10d\",18) = \"%10d\"",18);nr(1);
	printf("sprintf(\"%%-10d\",18) = \"%-10d\"",18);nr(1);

	printLn2(1,"組合せ",1);
	printf("sprintf(\"%%d年%%d月%%d日\",2007,7,14) = \"%d年%d月%d日\"",2007,7,14);nr(1);
	printf("sprintf(\"x = %%+08.2f\",0.375)      = \"x = %+08.2f\"",0.375);nr(1);

	printLn2(1,"引数並び",1);
	printf("sprintf(\"%%3$d年%%1$d月%%2$d日\",7,14,2007) = \"%3$d年%1$d月%2$d日\"",7,14,2007);nr(1);
	printf("sprintf(\"%%2$sは%%2$s。%%1$sは%%1$s。\",\"あれ\",\"これ\") = \"%2$sは%2$s。%1$sは%1$s。\"","あれ","これ");nr(1);

	nr(2);

}
/*
	%% とすることで % という文字として扱われる

	• printf("フォーマットテキスト",args)
		引数argsをフォーマットテキストにしたがってフォーマットして出力する
	• sprintf(str,"フォーマットテキスト",args)
		引数argsをフォーマットテキストにしたがってフォーマットし,文字列str (char *)に書き込まれる
*/